// react-native-integrate/lib/tasks/notificationServiceTask.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.summary = exports.runTask = exports.notificationServiceTask = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const constants_1 = require("../constants");
const applyContentModification_1 = require("../utils/applyContentModification");
const findClosingTagIndex_1 = require("../utils/findClosingTagIndex");
const getErrMessage_1 = require("../utils/getErrMessage");
const getProjectPath_1 = require("../utils/getProjectPath");
const satisfies_1 = require("../utils/satisfies");
const setState_1 = require("../utils/setState");
const stringSplice_1 = require("../utils/stringSplice");
const variables_1 = require("../variables");
async function notificationServiceTask(args) {
    let { content } = args;
    const { task, configPath, packageName } = args;
    for (const action of task.actions) {
        variables_1.variables.set('CONTENT', content);
        if (action.when && !(0, satisfies_1.satisfies)(variables_1.variables.getStore(), action.when)) {
            (0, setState_1.setState)(action.name, {
                state: 'skipped',
                reason: 'when',
                error: false,
            });
            continue;
        }
        (0, setState_1.setState)(action.name, {
            state: 'progress',
            error: false,
        });
        try {
            content = await (0, applyContentModification_1.applyContentModification)({
                action,
                findOrCreateBlock,
                configPath,
                packageName,
                content,
                indentation: 4,
            });
            (0, setState_1.setState)(action.name, {
                state: 'done',
                error: false,
            });
        }
        catch (e) {
            (0, setState_1.setState)(action.name, {
                state: 'error',
                reason: (0, getErrMessage_1.getErrMessage)(e),
                error: true,
            });
            throw e;
        }
    }
    return content;
}
exports.notificationServiceTask = notificationServiceTask;
function findOrCreateBlock(content, block) {
    let blockContent = {
        start: 0,
        end: content.length,
        match: content,
        space: '',
        justCreated: false,
    };
    const blockDefinition = blockDefinitions[block];
    if (!blockDefinition)
        throw new Error(`Invalid block: ${block}`);
    const { regex, makeNewMethod } = blockDefinition;
    let blockStart = regex.exec(content);
    const justCreated = !blockStart;
    if (!blockStart) {
        const newMethod = makeNewMethod();
        content = appendNewMethod(content, newMethod);
        blockStart = regex.exec(content);
    }
    if (!blockStart) {
        throw new Error('block could not be inserted, something wrong?');
    }
    const blockEndIndex = (0, findClosingTagIndex_1.findClosingTagIndex)(content, blockStart.index + blockStart[0].length);
    const blockBody = content.substring(blockStart.index + blockStart[0].length, blockEndIndex);
    blockContent = {
        start: blockStart.index + blockStart[0].length,
        end: blockEndIndex,
        match: blockBody,
        justCreated,
        space: '',
    };
    return {
        blockContent,
        content,
    };
}
const blockDefinitions = {
    didReceiveNotificationRequest: {
        regex: /didReceiveNotificationRequest:.*?withContentHandler.*?\{/s,
        makeNewMethod: () => {
            return '- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {}';
        },
    },
    serviceExtensionTimeWillExpire: {
        regex: /serviceExtensionTimeWillExpire.*?\{/s,
        makeNewMethod: () => {
            return '- (void)serviceExtensionTimeWillExpire {}';
        },
    },
};
function appendNewMethod(content, newMethod) {
    const notificationServiceMatch = /@implementation NotificationService.*?@end/s.exec(content);
    if (!notificationServiceMatch)
        throw new Error('Could not find @implementation NotificationService');
    const codeToInsert = `${newMethod}

`;
    return (0, stringSplice_1.stringSplice)(content, notificationServiceMatch.index + notificationServiceMatch[0].length - 4, 0, codeToInsert);
}
function getNotificationServicePath(target) {
    const projectPath = (0, getProjectPath_1.getProjectPath)();
    const notificationServicePath = path_1.default.join(projectPath, 'ios', target, constants_1.Constants.NOTIFICATION_SERVICE_FILE_NAME);
    if (!fs_1.default.existsSync(notificationServicePath))
        throw new Error(`NotificationService file not found at ${notificationServicePath}`);
    return notificationServicePath;
}
function readNotificationServiceContent(target) {
    const notificationServicePath = getNotificationServicePath(target);
    return fs_1.default.readFileSync(notificationServicePath, 'utf-8');
}
function writeNotificationServiceContent(content, target) {
    const notificationServicePath = getNotificationServicePath(target);
    return fs_1.default.writeFileSync(notificationServicePath, content, 'utf-8');
}
async function runTask(args) {
    args.task.target = (0, variables_1.getText)(args.task.target);
    let content = readNotificationServiceContent(args.task.target);
    content = await notificationServiceTask({
        ...args,
        content,
    });
    writeNotificationServiceContent(content, args.task.target);
}
exports.runTask = runTask;
exports.summary = 'NotificationService.m modification';
