// react-native-integrate/lib/tasks/promptTask.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isSystemTask = exports.summary = exports.runTask = exports.promptTask = void 0;
const getErrMessage_1 = require("../utils/getErrMessage");
const runPrompt_1 = require("../utils/runPrompt");
const satisfies_1 = require("../utils/satisfies");
const setState_1 = require("../utils/setState");
const variables_1 = require("../variables");
async function promptTask(args) {
    const { task, packageName } = args;
    for (const action of task.actions) {
        if (action.when && !(0, satisfies_1.satisfies)(variables_1.variables.getStore(), action.when)) {
            (0, setState_1.setState)(action.name, {
                state: 'skipped',
                reason: 'when',
                error: false,
            });
            continue;
        }
        (0, setState_1.setState)(action.name, {
            state: 'progress',
            error: false,
        });
        try {
            await (0, runPrompt_1.runPrompt)(action, packageName);
        }
        catch (e) {
            (0, setState_1.setState)(action.name, {
                state: 'error',
                reason: (0, getErrMessage_1.getErrMessage)(e),
                error: true,
            });
            throw e;
        }
    }
}
exports.promptTask = promptTask;
exports.runTask = promptTask;
exports.summary = '';
// noinspection JSUnusedGlobalSymbols
exports.isSystemTask = true;
