// react-native-integrate/lib/__tests__/unit/tasks/promptTask.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const promptTask_1 = require("../../../tasks/promptTask");
const variables_1 = require("../../../variables");
const mock = jest.spyOn(require('../../../utils/runPrompt'), 'runPrompt');
describe('promptTask', () => {
    it('should handle unexpected error', async () => {
        mock.mockImplementationOnce(() => {
            throw new Error('unexpected error');
        });
        const task = {
            task: 'prompt',
            actions: [
                {
                    name: 'test',
                    text: 'test',
                    type: 'text',
                },
            ],
        };
        await expect((0, promptTask_1.promptTask)({
            configPath: 'path/to/config',
            task: task,
            packageName: 'test-package',
        })).rejects.toThrowError('unexpected error');
        expect(variables_1.variables.get('test')).toEqual('error');
        mock.mockReset();
    });
});
